### PR TITLE
Refactor Push and Pull CLI commands

### LINKED ISSUE #3532: Refactor Pull and Push CLI Code
As brought up in #3526:
Currently the code in `/cmd/internal/cli/pull.go` and `/cmd/internal/cli/push.go` contains very long switch case statements that make it hard to follow. These individual case statements should be broken into functions in order to make them more readable and testable. These functions should live in `/internal/app/singularity`.

### PR DESCRIPTION
**Pull request by ikaneshiro**
**Description of the Pull Request (PR):**

- Takes long cases from switch statements in cli for `push` and `pull` commands and moves them into `internal/app/singularity` to make them easier to read and test in the future.

**This fixes or addresses the following GitHub issues:**

- Fixes #3532 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: cmd/internal/cli/pull.go [hunk 1]
```diff
@@ -223,160 +206,46 @@ func pullRun(cmd *cobra.Command, args []string) {
...
-			}
-		} else {
-			sylog.Warningf("Skipping container verification")
+		err := singularity.LibraryPull(name, ref, transport, args[i], PullLibraryURI, KeyServerURL, authToken, force, unauthenticatedPull)
+		if err == singularity.ErrLibraryUnsigned {
+			exitStat = 1
+		} else if err == singularity.ErrLibraryPullAbort {
+			exitStat = 10
+		} else if err != nil {
+			sylog.Fatalf("%v", err)
```
Comment (WestleyK) [inline cmd/internal/cli/pull.go]: Does this need a message? Like:

```suggestion
			sylog.Fatalf("Unable to pull contianer: %v", err)
```

File: cmd/internal/cli/pull.go [hunk 2]
```diff
@@ -223,160 +206,46 @@ func pullRun(cmd *cobra.Command, args []string) {
...
-			if desc.MediaType == SifLayerMediaType {
-				nameOld, _ := content.ResolveName(desc)
-				_ = store.MapPath(nameOld, name)
-			}
-			return nil, nil
-		}
-		pullHandler := oras.WithPullBaseHandler(images.HandlerFunc(handlerFunc))
-
-		_, _, err = oras.Pull(orasctx.Background(), resolver, spec.String(), store, allowedMediaTypes, pullHandler)
+		err = singularity.OrasPull(name, ref, force, ociAuth)
```
Comment (WestleyK) [inline cmd/internal/cli/pull.go:230]: If go will let you...

```suggestion
		err := singularity.OrasPull(name, ref, force, ociAuth)
```

File: cmd/internal/cli/pull.go [hunk 3]
```diff
@@ -223,160 +206,46 @@ func pullRun(cmd *cobra.Command, args []string) {
...
 	case ociclient.IsSupported(transport):
-		downloadOciImage(name, args[i], cmd)
+		ociAuth, err := makeDockerCredentials(cmd)
+		if err != nil {
+			sylog.Fatalf("While creating Docker credentials: %v", err)
+		}
+
+		err = singularity.OciPull(name, args[i], tmpDir, ociAuth, force, noHTTPS)
+		if err != nil {
+			sylog.Fatalf("%v", err)
```
Comment (WestleyK) [inline cmd/internal/cli/pull.go]: Maybe here too?

```suggestion
			sylog.Fatalf("Unable to pull oci image: %v", err)
```

File: cmd/internal/cli/push.go [hunk 4]
```diff
@@ -93,89 +74,22 @@ var PushCmd = &cobra.Command{
...
-
-			if err := client.UploadImage(file, dest, PushLibraryURI, authToken, "No Description"); err != nil {
-				sylog.Fatalf("%v\n", err)
+			err := singularity.LibraryPush(file, dest, authToken, PushLibraryURI, KeyServerURL, remoteWarning, unauthenticatedPush)
+			if err == singularity.ErrLibraryUnsigned {
+				fmt.Fprintf(os.Stderr, "\nUnable to verify your container! You REALLY should sign your container before pushing!\n")
+				fmt.Fprintf(os.Stderr, "Stopping upload.\n")
+				os.Exit(3)
+			} else if err != nil {
+				sylog.Fatalf("%v", err)
```
Comment (WestleyK) [inline cmd/internal/cli/push.go]:
```suggestion
				sylog.Fatalf("Unable to push image: %v", err)
```

File: cmd/internal/cli/push.go [hunk 5]
```diff
@@ -93,89 +74,22 @@ var PushCmd = &cobra.Command{
...
-			if err != nil {
-				sylog.Fatalf("Unable to add SIF file to FileStore: %s", err)
-			}
-
-			descriptors := []ocispec.Descriptor{desc}
-
-			if _, err := oras.Push(context.Background(), resolver, spec.String(), store, descriptors, oras.WithConfig(conf)); err != nil {
-				sylog.Fatalf("Unable to push: %s", err)
+			if err := singularity.OrasPush(file, ref, ociAuth); err != nil {
+				sylog.Fatalf("%v", err)
```
Comment (WestleyK) [inline cmd/internal/cli/push.go]: Here too?

```suggestion
				sylog.Fatalf("Unable to push oras image: %v", err)
```

File: internal/app/singularity/pull.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+	if !exists {
+		sylog.Infof("Downloading library image")
+		if err = client.DownloadImage(imagePath, fullURI, libraryURI, true, authToken); err != nil {
+			return fmt.Errorf("unable to Download Image: %v", err)
+		}
+
+		if cacheFileHash, err := client.ImageHash(imagePath); err != nil {
+			return fmt.Errorf("Error getting ImageHash: %v", err)
+		} else if cacheFileHash != libraryImage.Hash {
+			return fmt.Errorf("Cached File Hash(%s) and Expected Hash(%s) does not match", cacheFileHash, libraryImage.Hash)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
			return fmt.Errorf("cached file hash(%s) and expected hash(%s) does not match", cacheFileHash, libraryImage.Hash)
￼
```

File: internal/app/singularity/pull.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+		return fmt.Errorf("unable to check if %v exists: %v", imagePath, err)
+	}
+	if !exists {
+		sylog.Infof("Downloading library image")
+		if err = client.DownloadImage(imagePath, fullURI, libraryURI, true, authToken); err != nil {
+			return fmt.Errorf("unable to Download Image: %v", err)
+		}
+
+		if cacheFileHash, err := client.ImageHash(imagePath); err != nil {
+			return fmt.Errorf("Error getting ImageHash: %v", err)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
			return fmt.Errorf("error getting imageHash: %v", err)
```

File: internal/app/singularity/pull.go [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+			fmt.Fprintf(os.Stderr, "This image is not signed, and thus its contents cannot be verified.\n")
+			resp, err := sypgp.AskQuestion("Do you with to proceed? [N/y] ")
+			if err != nil {
+				return fmt.Errorf("unable to parse input: %v", err)
+			}
+			if resp == "" || resp != "y" && resp != "Y" {
+				fmt.Fprintf(os.Stderr, "Aborting.\n")
+				err := os.Remove(name)
+				if err != nil {
+					return fmt.Errorf("unabel to delete the container: %v", err)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
					return fmt.Errorf("unable to delete the container: %v", err)
```

File: internal/app/singularity/pull.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+// OrasPull will download the image specified by the provided oci reference and store
+// it at the location specified by file, it will use credentials if supplied
+func OrasPull(name, ref string, force bool, ociAuth *ocitypes.DockerAuthConfig) error {
+	ref = strings.TrimPrefix(ref, "//")
+
+	spec, err := reference.Parse(ref)
+	if err != nil {
+		return fmt.Errorf("Unable to parse oci reference: %s", err)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
		return fmt.Errorf("unable to parse oci reference: %s", err)
```

File: internal/app/singularity/pull.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	credFn := func(_ string) (string, string, error) {
+		return ociAuth.Username, ociAuth.Password, nil
+	}
+	resolver := docker.NewResolver(docker.ResolverOptions{Credentials: credFn})
+
+	wd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("Failed to get working directory: %s", err)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
		return fmt.Errorf("failed to get working directory: %s", err)
```

File: internal/app/singularity/pull.go [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+	sylog.Infof("Download complete: %s\n", name)
+
+	return nil
+}
+
+// OciPull will build a SIF image from the specified oci URI
+func OciPull(name, imageURI, tmpDir string, ociAuth *ocitypes.DockerAuthConfig, force, noHTTPS bool) error {
+	if !force {
+		if _, err := os.Stat(name); err == nil {
+			return fmt.Errorf("Image file already exists - will not overwrite")
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
			return fmt.Errorf("image file already exists - will not overwrite")
```

File: internal/app/singularity/pull.go [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+			Opts: types.Options{
+				TmpDir:           tmpDir,
+				NoTest:           true,
+				NoHTTPS:          noHTTPS,
+				DockerAuthConfig: authConf,
+			},
+		},
+	)
+	if err != nil {
+		return fmt.Errorf("Unable to create new build: %v", err)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
		return fmt.Errorf("unable to create new build: %v", err)
```

**Review by WestleyK — CHANGES_REQUESTED** [review]
Couple small suggestions, otherwise looks good.

File: cmd/internal/cli/pull.go (hunk 2 above)
Comment (ikaneshiro) [reply cmd/internal/cli/pull.go:230]: `err` already exists in this scope, I could embed these calls in an `if` statement to limit the scope, but that would mean some pretty long lines

File: internal/app/singularity/pull.go [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+		}
+
+		if cacheFileHash, err := client.ImageHash(imagePath); err != nil {
+			return fmt.Errorf("error getting imagehash: %v", err)
+		} else if cacheFileHash != libraryImage.Hash {
+			return fmt.Errorf("cached file hash(%s) and expected hash(%s) does not match", cacheFileHash, libraryImage.Hash)
+		}
+	}
+
+	// Perms are 777 *prior* to umask
```
Comment (mem) [inline internal/app/singularity/pull.go]: The comment should refer to the unusual choice of 0777 for what should be a regular file, noting that the SIF file is created with a shebang to be able to run the image as a "script". I believe this comment was the in the original version.

File: internal/app/singularity/pull.go [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+	dstFile, err := os.OpenFile(name, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+	if err != nil {
+		return fmt.Errorf("%v", err)
+	}
+	defer dstFile.Close()
+
+	srcFile, err := os.OpenFile(imagePath, os.O_RDONLY, 0444)
```
Comment (mem) [inline internal/app/singularity/pull.go]: again, I believe this was here before, but the mode argument does not apply in this case.

**Review by mem — COMMENTED** [review]
LGTM, I'll let @WestleyK decide

File: internal/app/singularity/pull.go [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+		if err != nil {
+			// err will be: "unable to verify container: %v", err
+			sylog.Warningf("%v", err)
+			// if theres a warning, return set error to indicate exit 1
+			retErr = ErrLibraryUnsigned
+		}
+		// if container is not signed, print a warning
+		if !imageSigned {
+			fmt.Fprintf(os.Stderr, "This image is not signed, and thus its contents cannot be verified.\n")
+			resp, err := sypgp.AskQuestion("Do you with to proceed? [N/y] ")
```
Comment (gvallee) [inline internal/app/singularity/pull.go]:
```suggestion
			resp, err := sypgp.AskQuestion("Do you want to proceed? [N/y] ")
```

File: internal/app/singularity/pull.go [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	var retErr error
+	// check if we pulled from the library, if so; is it signed?
+	if !unauthenticated {
+		imageSigned, err := signing.IsSigned(name, keyServerURL, 0, false, authToken, true)
+		if err != nil {
+			// err will be: "unable to verify container: %v", err
+			sylog.Warningf("%v", err)
+			// if theres a warning, return set error to indicate exit 1
```
Comment (gvallee) [inline internal/app/singularity/pull.go]:
```suggestion
			// if there is a warning, return set error to indicate exit 1
```

File: internal/app/singularity/pull.go [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+			if err != nil {
+				return fmt.Errorf("unable to parse input: %v", err)
+			}
+			if resp == "" || resp != "y" && resp != "Y" {
+				fmt.Fprintf(os.Stderr, "Aborting.\n")
+				err := os.Remove(name)
+				if err != nil {
+					return fmt.Errorf("unable to delete the container: %v", err)
+				}
+				// set error to exit with status 10 after replying no
```
Comment (gvallee) [inline internal/app/singularity/pull.go]: Would it be possible to improve the comment? I personally do not understand what it really means.

File: internal/app/singularity/push.go [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	if err := client.UploadImage(file, dest, libraryURI, authToken, "No Description"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// OrasPush upload the image specified by file and push it to the provided oci reference,
```
Comment (gvallee) [inline internal/app/singularity/push.go]:
```suggestion
// OrasPush uploads the image specified by file and pushes it to the provided oci reference,
```

**Review by gvallee — DISMISSED** [review]
I only did a quick read-through and marked a few quick points to fix. I can check the semantics if required, just let me know.

File: cmd/internal/cli/pull.go [hunk 19]
```diff
@@ -223,160 +206,46 @@ func pullRun(cmd *cobra.Command, args []string) {
...
-			}
-		} else {
-			sylog.Warningf("Skipping container verification")
-		}
-		fmt.Printf("Download complete: %s\n", name)
-
 	case ShubProtocol:
 		err := shub.DownloadImage(name, args[i], force, noHTTPS)
 		if err != nil {
 			sylog.Fatalf("%v\n", err)
```
Comment (WestleyK) [inline cmd/internal/cli/pull.go]: Does this need:

```suggestion
 			sylog.Fatalf("Unable to download shub image: %v", err)
```

File: cmd/internal/cli/pull.go [hunk 20]
```diff
@@ -223,160 +206,46 @@ func pullRun(cmd *cobra.Command, args []string) {
...
-		// ensure container is executable
-		if err := os.Chmod(name, 0755); err != nil {
-			sylog.Fatalf("Unable to set image perms: %s", err)
+			sylog.Fatalf("While pulling image from oci registry: %v", err)
 		}
-		sylog.Infof("Download complete: %s\n", name)
 	case HTTPProtocol, HTTPSProtocol:
 		err := net.DownloadImage(name, args[i], force)
 		if err != nil {
 			sylog.Fatalf("%v\n", err)
```
Comment (WestleyK) [inline cmd/internal/cli/pull.go]: Here too?

```suggestion
 			sylog.Fatalf("Unable to ...: %v\n", err)
```

File: internal/app/singularity/pull.go [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+			return fmt.Errorf("error getting imagehash: %v", err)
+		} else if cacheFileHash != libraryImage.Hash {
+			return fmt.Errorf("cached file hash(%s) and expected hash(%s) does not match", cacheFileHash, libraryImage.Hash)
+		}
+	}
+
+	// Perms are 777 *prior* to umask
+	dstFile, err := os.OpenFile(name, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+	if err != nil {
+		return fmt.Errorf("%v", err)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]: Can you just return the err? Instead of passing it thrught `fmt.Errorf()`?

```suggestion
		return err
```

File: internal/app/singularity/pull.go [hunk 22]
```diff
@@ -0,0 +1,269 @@
...
+	sylog.Infof("Download complete: %s\n", name)
+
+	return nil
+}
+
+// OciPull will build a SIF image from the specified oci URI
+func OciPull(name, imageURI, tmpDir string, ociAuth *ocitypes.DockerAuthConfig, force, noHTTPS bool) error {
+	if !force {
+		if _, err := os.Stat(name); err == nil {
+			return fmt.Errorf("image file already exists - will not overwrite")
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]: Maybe add:

```suggestion
			return fmt.Errorf("image file already exists: %q - will not overwrite", name)
```

**Review by WestleyK — CHANGES_REQUESTED** [review]
Couple more suggestions :+1:

File: internal/app/singularity/pull.go [hunk 23]
```diff
@@ -0,0 +1,270 @@
...
+	// executed with its leading shebang like a script
+	dstFile, err := os.OpenFile(name, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0777)
+	if err != nil {
+		return err
+	}
+	defer dstFile.Close()
+
+	srcFile, err := os.Open(imagePath)
+	if err != nil {
+		return fmt.Errorf("%v", err)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]:
```suggestion
		return err
```
Or:

```suggestion
		return fmt.Errorf("unable to open file: %v", err)
```

File: internal/app/singularity/pull.go [hunk 24]
```diff
@@ -0,0 +1,270 @@
...
+	}
+	defer srcFile.Close()
+
+	// Copy SIF from cache
+	_, err = io.Copy(dstFile, srcFile)
+	if err != nil {
+		return fmt.Errorf("%v", err)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]: Same here.

File: internal/app/singularity/pull.go [hunk 25]
```diff
@@ -0,0 +1,270 @@
...
+	cachedImgPath := cache.OciTempImage(sum, imgName)
+
+	exists, err := cache.OciTempExists(sum, imgName)
+	if err != nil {
+		return fmt.Errorf("unable to check if %s exists: %s", imgName, err)
+	}
+	if !exists {
+		sylog.Infof("Converting OCI blobs to SIF format")
+		if err := convertDockerToSIF(imageURI, cachedImgPath, tmpDir, noHTTPS, ociAuth); err != nil {
+			return fmt.Errorf("%v", err)
```
Comment (WestleyK) [inline internal/app/singularity/pull.go]: Same here.

**Review by WestleyK — CHANGES_REQUESTED** [review]
Couple more suggestions...

**Review by WestleyK — APPROVED** [review]
Looks good to me :+1:

**Comment by ikaneshiro** [comment]
@WestleyK @gvallee going to wait to merge this until #3578 is merged in order to work the conflicts into this refactoring

**Comment by ikaneshiro** [comment]
@WestleyK the merge required a lot of manual conflict resolution, can you give this a quick look to see if I messed anything up?

**Review by WestleyK — APPROVED** [review]
Yup, still looks good to me :+1:
